### ISSUE TITLE
Blank page on output from pdfsizeopt

### ISSUE THREAD

**Issue by rbrito** [pts/pdfsizeopt#30]
Dear @pts,

I don't know if the file linked below is corrupt or not, but it shows a behavior (with fonts) that is funny: its page 4 (if I am not mistaken) is completely blank after being processed with pdfsizeopt. The file (and its variant) are the following:

http://nlp.stanford.edu/IR-book/pdf/irbookonlinereading.pdf
http://nlp.stanford.edu/IR-book/pdf/irbookprint.pdf

I hope that you are able to reproduce the problem on your side.

Thanks,

Rogério.

**Comment by pts** [comment]
I'm not able to reproduce this problem.

Please specify which file, exactly which page, and attach screenshots before and after conversion with pdfsizeopt.

**Comment by rbrito** [comment]
The page in question is page 4.

I just took the first link (irbookonlinereading.pdf) to show the problem. I run pdfsizeopt on the whole file. I extracted both pages 4 from the input, but pdftk refused extract only page 4 from the entire output of pdfsizeopt:

```
$ pdftk irbookonlinereading.pdf cat 4 output p4.pdf
$ pdftk irbookonlinereading.pso.pdf cat 4 output output.pdf
Error: Unable to find file.
Error: Failed to open PDF file: 
   irbookonlinereading.pso.pdf
Errors encountered.  No output created.
Done.  Input errors, so no output created.
```

Interestingly, if I run pdfsizeopt only on the extracted page, the output page is OK.

I am attaching:

* the screenshot with the input and output file from pdfsizeopt.
* only page 4 of the original document

![before-and-after-pdfsizeopt](https://user-images.githubusercontent.com/212145/29794747-b509d4fa-8c1f-11e7-90f5-0855299d055f.png)
[p4.pdf](https://github.com/pts/pdfsizeopt/files/1258445/p4.pdf)

I will try more things with the file, though.

**Comment by rbrito** [comment]
Oh, right there at the beginning of pdfsizeopt's output while processing the original document, there is a warning generated:

```
$ ~/Downloads/pdfsizeopt/pdfsizeopt --use-pngout=no --use-multivalent=no irbookonlinereading.pdf 
info: This is pdfsizeopt rUNKNOWN size=353887.
info: prepending to PATH: /home/rbrito/Downloads/pdfsizeopt
info: loading PDF from: irbookonlinereading.pdf
info: loaded PDF of 6903344 bytes
info: separated to 12786 objs + xref + trailer
warning: cannot parse obj 3293: pdfsizeopt.main.PdfTokenParseError: X Y obj expected, got 'K\xedugQy&\x04C\xe1l\xb4QV!\x04\xa9\xc0\xe6D\xdb\x9fx\xf1T\xcbi\xb4\x8ahK\x1a' at ofs=2397
warning: cannot parse obj 277: pdfsizeopt.main.PdfTokenParseError: expected endstream+endobj in obj 277 at 2849
info: found 0 Type1 fonts loaded
info: found 44 Type1C fonts loaded
info: using Ghostscript gs: GPL Ghostscript 9.21 (2017-03-16)
info: decompressing 16244 bytes with Ghostscript /Filter/LZWDecode
(...)
```

Note the problems parsing the file right there at the beginning.

**Comment by pts** [comment]
Thank you for reporting this bug, and thank you for providing additional details! I was able to reproduce the problem based on your instructions.

Indeed, page 4 of irbookonlinereading.pso.pdf is blank (even when viewed in Ghostscript). However, the page of p4.pso.pdf is good.

Indeed, I'm also getting the following warnings, and they seem to be relevant:

```
warning: cannot parse obj 3293: pdfsizeopt.main.PdfTokenParseError: X Y obj expected, got 'K\xedugQy&\x04C\xe1l\xb4QV!\x04\xa9\xc0\xe6D\xdb\x9fx\xf1T\xcbi\xb4\x8ahK\x1a' at ofs=2397
warning: cannot parse obj 277: pdfsizeopt.main.PdfTokenParseError: expected endstream+endobj in obj 277 at 2849
```

This is not an image or font optimization bug, because it also happens with `pdfsizeopt --do-optimize-images=no --do-optimize-fonts=no irbookonlinereading.pdf`.

The PDF file irbookonlinereading.pdf is broken. It has the xref entry `0000002397 00000 n` for obj 3293, but there is no obj there in the file. In fact, offset 2397 is in the middle of obj 277. obj 277 in itself is  correct, but the situation (that xref entry of another obj points into the middle of obj 277) confuses pdfsizeopt, and it fails to parse obj 277, and this causes page 4 to be blank.

I'll prepare a fix for pdfsizeopt to remove this confusion. This would make obj 277 work, which would fix page 4. We don't care about the broken and missing obj 3293, because it's not used (there is no `3239 0 R`).

**Comment by pts** [comment]
Fixed in c7af2a1c6505ad4ef8adee291286e365226393c4.

Thank you for reporting this bug!
